Fix bug in analyzing preprocessor arguments

webmaster128 opened the pull request.

frerich commented on unittests.py:
```diff
@@ -183,6 +186,32 @@ def testOutputFileFromSourcefile(self):
+        self._testPreprocessingOutfile(['/c', '/P', '/EP', 'main.cpp'], None)
+        self._testPreprocessingOutfile(['/c', '/P', '/EP', '/FoSome.obj', 'main.cpp'], None)
+        self._testPreprocessingOutfile(['/c', '/P', '/EP', '/FaListing.asm', 'main.cpp'], None)
+        self._testPreprocessingOutfile(['/c', '/P', '/EP', '/FmMapfile.map', 'main.cpp'], None)
```
I'm not sure these last four tests are correct, for me (using Visual Studio 2013), passing both `/P` and `/EP` makes the compiler write the preprocessed output to an `.i` file, too (i.e. `/P` overrules `/EP`). Note that this contradicts the documentation [which says](https://msdn.microsoft.com/en-us/library/becb7sys(v=vs.120).aspx)

> To send the preprocessed output to stdout, with #line directives, use /P and /EP together.

webmaster128 replied:
Nice one, thanks! I have the same behavior using MSVS2015. /P is with #line annotations and /P /EP is without – both go to file.

frerich commented on clcache.py:
```diff
@@ -817,29 +817,27 @@ def analyze(cmdline):
-            else:
-                # Preprocess to stdout. Use empty string rather then None to ease
-                # output to log.
-                outputFile = ''
```
Hm, I see commit a3771033 intentionally did not use `None` here, but I don't quite understand the comment. I suspect there was some string concatenation or so going on which `None` prohibits? In any case, this doesn't seem to hurt... I'm just wary of reverting code which has a comment attached to it explaining that things are done intentionally.

webmaster128 replied:
Ah thanks for looking at the source. [When vchigrin commited this](https://github.com/frerich/clcache/blob/a37710335e62e98a3140757743fa161f2d818057/clcache.py), we introduced the first non-path outputFile value and the rest of the code was not ready to deal with it. [Existing](https://github.com/frerich/clcache/blob/a37710335e62e98a3140757743fa161f2d818057/clcache.py#L777) [tracing](https://github.com/frerich/clcache/blob/a37710335e62e98a3140757743fa161f2d818057/clcache.py#L791) and [os path](https://github.com/frerich/clcache/blob/a37710335e62e98a3140757743fa161f2d818057/clcache.py#L779) [methods](https://github.com/frerich/clcache/blob/a37710335e62e98a3140757743fa161f2d818057/clcache.py#L790) could not deal with `None`. So he did a plausible thing in order to get a minimal commit.

frerich commented:
Thanks, looks good except for the part where the tests verify the documented instead of the actual (at least for me) behaviour of the compiler.

I guess eventually we'll want to support a list of output files in order to implement support for #30.

webmaster128 commented:
Thanks for review. I think I'll add a small helper function `_preprocessToStdout` in which I document the documentation bug.

Yes, #30 is one big goal. Beeing unable to break existing functionality by excessive testing is another.

webmaster128 commented:
Check

* cleaned up everything
* changed test cases, implementation and changelog for the /E /EP case
* pulled out preprocessToStdout that explicitly takes care of the documentation bug

Next up: test more different /Fi values, detect bug when a directory is used, fix implementation, handle /Fo and /Fi using a common helper.

frerich commented on clcache.py:
```diff
@@ -737,6 +737,21 @@ def expandCommandLine(cmdline):
+        if 'P' in options:
+            return False
+        else:
+            return True
```
Is the `assert` maybe a bit harsh? I think Python permits expressing whether some options cause preprocessor output to be printed to `stdout` in almost plain english:

```python
def _preprocessToStdout(options):
    return ('E' in options or 'EP' in options) and 'P' not in options
```

I think that reads very nicely.

webmaster128 replied:
My rule of thumb is to place assertions whenever a failing expression is a programming error.

This method must never be called when not in preprocessing because the entire logic only woks there. So instead of pushing the error down the road where it gets harder to debug, we should fail fast.

I am fine with rewriting the implementation in one line but I think the assertion adds value.

frerich replied:
I agree on the value of assertions in general, but for a method call `_preprocessToStdout` I think it's perfectly fine if it yields `True` if the given options indicate that preprocessed output will be printed to `stdout` and `False` otherwise. I.e. I think it's plausible that this method returns `False` for e.g. empty lists.

webmaster128 replied:
ok

frerich commented:
Looks fine to me except for the somewhat elaborate definition of `_preprocessToStdout`.

webmaster128 commented:
Alright, alternative implementation of  preprocessToStdout() is in

frerich commented:
Thanks, looks excellent - merged!
